### PR TITLE
Migrate Header Component to use React Hooks

### PR DESCRIPTION
**Pull request by ntsummers1**
### Description

As part of an larger dependency upgrade to React-Router, I'll be updating code slowly overtime to use React hooks where possible throughout the code. This is because React-Redux and React-Router are both pushing towards using hooks and will stop or have stopped supporting older methods. Thus, to make this transition easier and to make sure we're on keeping our code up to date, I'll be going through where possible to make these updates.

This first PR updates the Header component and does the following:

1. Updates the component from using a class structure to a functional component structure
2. Updates the react-router-dom call to the location to use the associated hook
3. Updates the react-redux state calls to use the useSelector hook

---
### How to test

Run the application
Play around with the header. See that you can navigate to the profile page, it adds the autosave subheader when on a report, it updates the autosave header when updating a report, and hides it if the report is certified and submitted.

Also, click around with the dropdown open. Make sure it closes if you click on the background, closes if you navigate to a different page, closes if you click on the dropdown itself. Also that it can reopen after you close it.

---
### Pre-review checklist

- [x] I have added [thorough](https://shorturl.at/aejkF) tests, if necessary
- [x] I have updated relevant documentation, if necessary
- [x] I have performed a self-review of my code
- [x] I have manually tested this PR in the deployed cloud environment

---
#### Review
- [x] Design: This work has been reviewed and approved by design, if necessary
- [x] Product: This work has been reviewed and approved by product owner, if necessary

### CODE REVIEW

**Review by gmrabian — DISMISSED** [review]

**Review by karla-vm — DISMISSED** [review]
LGTM!

File: services/ui-src/src/components/layout/CertifyAndSubmit.jsx [hunk 1]
```diff
@@ -149,8 +149,20 @@ CertifyAndSubmit.defaultProps = {
 
 const mapState = (state) => ({
   isCertified: !selectIsFormEditable(state),
-  lastSave: moment(getCurrentReportStatus(state).lastChanged),
-  user: getCurrentReportStatus(state).username,
+  lastSave: moment(
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/layout/CertifyAndSubmit.jsx:152]: Suggested refactor for readability and calling `getCurrentReportStatus` once:

```
const mapState = (state) => {
  const { reportStatus, formData, stateUser, global } = state;

  const currentReport = getCurrentReportStatus(
    reportStatus,
    formData,
    stateUser,
    global
  );

  return {
    isCertified: !selectIsFormEditable(state),
    lastSave: moment(currentReport.lastChanged),
    user: currentReport.username,
    currentUserRole: stateUser.currentUser.role,
    state: stateUser.abbr,
  }
};
```

File: services/ui-src/src/components/layout/Header.jsx [hunk 2]
```diff
@@ -1,43 +1,47 @@
...
+  const [stateUser, formData, global, reportStatus] = useSelector(
+    (state) => [
+      state.stateUser,
+      state.formData,
+      state.global,
+      state.reportStatus,
+    ],
+    shallowEqual
+  );
+  const currentUser = stateUser.currentUser;
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/layout/Header.jsx:29]: Could use `const { currentUser } = stateUser;`

File: services/ui-src/src/components/layout/Header.jsx [hunk 3]
```diff
@@ -49,137 +53,110 @@ class Header extends Component {
...
-    const isLoggedIn = !!currentUser.username;
-    const { isMenuOpen } = this.state;
-    const shouldShowAutosave = this.showAutoSaveOnReport(
-      location,
-      currentUser,
-      reportStatus
-    );
+  useEffect(() => {
+    setIsMenuOpen(false);
+  }, [location?.pathname]);
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/components/layout/Header.jsx:64]: Is location ever undefined?

File: services/ui-src/src/store/selectors.js [hunk 4]
```diff
@@ -167,9 +173,14 @@ export const getCurrentReportStatus = (state) => {
  * @returns {boolean}
  */
 export const selectIsFormEditable = (state) => {
-  const { stateUser } = state;
+  const { reportStatus, formData, stateUser, global } = state;
   const { role } = stateUser.currentUser;
-  const status = getCurrentReportStatus(state).status;
+  const status = getCurrentReportStatus(
```
Comment (bangbay-bluetiger) [inline services/ui-src/src/store/selectors.js:178]: Could do this?
```
const { status } = getCurrentReportStatus(
  reportStatus,
  formData,
  stateUser,
  global
);
```

File: services/ui-src/src/components/layout/CertifyAndSubmit.jsx (hunk 1 above)
Comment (ntsummers1) [reply services/ui-src/src/components/layout/CertifyAndSubmit.jsx:152]: Will be doing this as part of the [Carts Pages component updates](https://jiraent.cms.gov/secure/RapidBoard.jspa?rapidView=5874&projectKey=CMDCT&view=planning&selectedIssue=CMDCT-3812&issueLimit=100#) :D

**Review by bangbay-bluetiger — DISMISSED** [review]

**Review by britt-mo — DISMISSED** [review]
Tested, everything worked as expected

**Review by bangbay-bluetiger — APPROVED** [review]

**Pull request merged**